### ISSUE TITLE
Adventure map, passability test

### ISSUE THREAD

**Issue by LeHerosInconnu** [ihhub/fheroes2#4495]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

Hello @ihhub,

So, I finished my big (and boring) test on passability without any object adjacency.
There are still many differences, here some examples:

Sawmills.
![Passability test 015](https://user-images.githubusercontent.com/43583994/138714812-55a54129-918b-420d-8870-b698d652948c.png)
![Passability test 016](https://user-images.githubusercontent.com/43583994/138714828-157dea7a-5162-483c-90c3-02967b5e33a7.png)
![Passability test 017](https://user-images.githubusercontent.com/43583994/138714843-bc5902dc-bb20-4ab0-b560-b2941a952d9c.png)
![Passability test 018](https://user-images.githubusercontent.com/43583994/138714853-8d35b789-5232-42a9-9a73-f699f8bcfb9c.png)
![Passability 02 013](https://user-images.githubusercontent.com/43583994/138714973-c34fc5c5-0e74-42af-b502-1c6985764c7c.png)
![Passability 02 014](https://user-images.githubusercontent.com/43583994/138714986-ad94bb5c-5a6b-4295-9e9a-112d52ce079f.png)
![Passability 02 015](https://user-images.githubusercontent.com/43583994/138715002-2bb6fb60-271e-4263-b808-73a731cf63ff.png)
![Passability 02 016](https://user-images.githubusercontent.com/43583994/138715009-60c08bb1-08fc-40cd-8957-d0541862c069.png)
![Passability 02 069](https://user-images.githubusercontent.com/43583994/138715194-45125f01-28c8-4f9a-b4fd-083745a8e6ae.png)
![Passability 02 070](https://user-images.githubusercontent.com/43583994/138715202-f5d12f7d-b8a0-4a93-89c8-b5fa5651cf97.png)

Water Lake.
![Passability 02 045](https://user-images.githubusercontent.com/43583994/138715245-955aa118-9e3f-4108-88ef-47327d9283b9.png)
![Passability 02 046](https://user-images.githubusercontent.com/43583994/138715256-5cefe5a3-04e2-4dfa-aa4f-9a2f265bc48f.png)
![Passability 02 047](https://user-images.githubusercontent.com/43583994/138715266-e99f718e-4589-4c09-b898-dac96f434e75.png)
Note that I had a problem while switching between the original game and fheroes2, the screen is a bit blurred and thus displayed in a different color, I didn't manage to reproduce this flaw.
![Passability 02 048](https://user-images.githubusercontent.com/43583994/138715282-b284aad3-9045-4540-9c0c-f64dd96a05eb.png)

There is also a difference of operating to activate the event of certain objects (https://github.com/ihhub/fheroes2/issues/4508).

In original game:
For Shipwreck and Magellan's Maps, the hero move on the active square of the object to activate the event.
And for Derelict Ship, Mermaid and Sirens, the hero don't move over the active square of the object to activate the event.

In fheroes2:
For Shipwreck, the hero don't move over the active square of the object to activate the event.
And for Derelict Ship, Mermaid, Sirens and Magellan's Maps, the hero move on the active square of the object to activate the event.

Tell me if you want me to open one or more issues for these differences.

There are also differences in passability with the Shipwreck and Derelict Ship.
And in the case of the Shipwreck, there is a difference in passability in the original game depending on the presence or absence of terrain south of the Shipwreck position.
I think that in the case of the Shipwreck and the presence of terrain south of the position it may be a question of a bug in the original game.

In original game:
![Passability test 069](https://user-images.githubusercontent.com/43583994/138716745-9c917a28-78cd-4d8e-8997-de36d7bd4e87.png)
![Passability test 106](https://user-images.githubusercontent.com/43583994/138716733-65830a9d-aad8-4300-bbac-a892c866e749.png)

Also, I think that all the squares of the Shipwreck and the Derelict Ship should remain accessible as it is currently the case in fheroes2, it is much more natural and it does not change at all the way to play compared to the original game.

In fheroes2:
![Passability test 066](https://user-images.githubusercontent.com/43583994/138717034-b524d5cf-6116-4e64-9fd1-85b6580e232d.png)
![Passability test 070](https://user-images.githubusercontent.com/43583994/138717022-13ba364a-19e5-45dd-b547-3a20f9c0e8b6.png)

There are also differences between the predetermined path of the hero, but these are not related to passability.
I will open another issue to optimize the predetermined path of the hero (https://github.com/ihhub/fheroes2/issues/4510).

![Passability test 001](https://user-images.githubusercontent.com/43583994/138723229-3ebf7625-5ef4-4255-ad1b-9e1792a4c4a6.png)
![Passability test 002](https://user-images.githubusercontent.com/43583994/138723248-b921d4a3-6df7-4037-841c-8f6c018ef2b5.png)

There are also differences with the way Whirlpools work (https://github.com/ihhub/fheroes2/issues/4509).
Related discussions: https://github.com/ihhub/fheroes2/discussions/3565, https://github.com/ihhub/fheroes2/discussions/3563 and https://github.com/ihhub/fheroes2/discussions/3562.

![Passability test 071](https://user-images.githubusercontent.com/43583994/138717299-1b206cd1-734c-4888-9118-47ceaa5f544d.png)
![Passability test 072](https://user-images.githubusercontent.com/43583994/138717310-5335f4be-e4a3-493c-92bc-f520996baa7a.png)

For the Jail, the hero passes through the Jail when he moves.
This bug is also present in the original game.
I think this can be fixed with the passability.
Tell me if you want me to open one separate issue for this.

![vlcsnap-2021-10-25-14h57m56s078](https://user-images.githubusercontent.com/43583994/138717547-15ecf174-a16e-4d22-8bd6-7eedab74f5b5.png)

https://user-images.githubusercontent.com/43583994/138717630-d651adf6-8448-4d20-a241-6b6fc7e80e65.mp4

Some objects don't have passability, but it's the same in the original game (the case of the Jail is mentioned above).
I think they can stay that way (except for the Jail).

![Passability 02 053](https://user-images.githubusercontent.com/43583994/138718143-c0bf65d2-85bf-40ee-b037-5d89b690537f.png)
![Passability 02 054](https://user-images.githubusercontent.com/43583994/138718159-d57b9b69-c840-4d30-9224-2c8a0b7d5002.png)
![Passability 02 081](https://user-images.githubusercontent.com/43583994/138718166-1aea947d-e6c1-4496-a1f6-ab82940126b6.png)

For me it's also okay (and preferable) that the Jail can be opened when the hero is located in the square directly north.
In the original game, the hero first moves east to open the Jail, which doesn't make much sense. :D

![Passability test 025](https://user-images.githubusercontent.com/43583994/138718890-a9ceccfd-8faa-4711-92e4-58f3502a972a.png)
![Passability test 026](https://user-images.githubusercontent.com/43583994/138718905-6ae6fcb3-67e3-479e-b625-6dff4ad3bc4a.png)

Now I just have to test all the possibilities with adjacency, see you in 10 years. :D
And then the passability with the Dimension Door spell, count another 10 years for that. :D :D
Note that this is a joke, I have no intention of doing this. :)

For passibility with the Dimension Door spell, a simple rule should be applied.
If the hero can move on a square on horseback, then he can also move on this square with the Dimension Door spell except for the interactive square(s) of the object (even for squares where the hero's path on horseback is blocked by another object(s) but which would be accessible if not blocked).

Here are all screenshots for passibility and predefined path differences:
https://drive.google.com/file/d/12eG_uOTrI7wSsnGJBagfvu8AGQ-qxAD0/view?usp=sharing

### Save file

Here are the test scenario and the save files:
[Passability test scenarios and saves.zip](https://github.com/ihhub/fheroes2/files/7411386/Passability.test.scenarios.and.saves.zip)

**Comment by Branikolog** [comment]
OG:
![image](https://user-images.githubusercontent.com/55348946/146356500-78f33934-3fa3-46e4-a235-9e6b82a20294.png)
fheroes2:
![image](https://user-images.githubusercontent.com/55348946/146356661-c7f9009e-3cb7-473a-bcf0-b31837b350e1.png)
[Pass.zip](https://github.com/ihhub/fheroes2/files/7726552/Pass.zip)

**Comment by ihhub** [comment]
The case described by @Branikolog is fixed. The only passability issue is left with Sawmill.

**Comment by Branikolog** [comment]
Sawmill in the OG while enemy hero is visiting it:
![sawmill](https://user-images.githubusercontent.com/55348946/152519758-ddc5a2f0-edca-4ab0-a125-f715e25b5c05.png)

**Comment by LeHerosInconnu** [comment]
Hello @Branikolog,

> Sawmill in the OG while enemy hero is visiting it: ![sawmill](https://user-images.githubusercontent.com/55348946/152519758-ddc5a2f0-edca-4ab0-a125-f715e25b5c05.png)

This is a bug in the original game that affects many objects and which should be fixed in fheroes2. :)
Link to some examples: https://heroes2.forumactif.com/t862-heroes-2-bugs-heroes-interaction-through-objects

**Comment by Branikolog** [comment]
Hi, @LeHerosInconnu !
Thanks for the info.
I've posted it here, because this can help our devs to fix current passability problems.

**Comment by Branikolog** [comment]
Fheroes2:
![image](https://user-images.githubusercontent.com/55348946/162442454-6aebc308-a596-4d9f-ba3b-1a6f76579e6d.png)
OG:
![image](https://user-images.githubusercontent.com/55348946/162442500-0d14d519-25d5-4a3d-882b-0bbd4c11413e.png)

**Comment by Branikolog** [comment]
One more object with wrong passability.
OG:
![image](https://user-images.githubusercontent.com/55348946/182028683-ec7d92ca-3d51-4682-9431-5f013de81f5d.png)
fheroes2:
![image](https://user-images.githubusercontent.com/55348946/182028703-6c1241df-abba-4020-8d1a-55df0269b8b8.png)

**Comment by ihhub** [comment]
> One more object with wrong passability. OG: ![image](https://user-images.githubusercontent.com/55348946/182028683-ec7d92ca-3d51-4682-9431-5f013de81f5d.png) fheroes2: ![image](https://user-images.githubusercontent.com/55348946/182028703-6c1241df-abba-4020-8d1a-55df0269b8b8.png)

Hi @Branikolog , I tested it on a new map and this object has correct passability.

**Comment by Branikolog** [comment]
Hello, @ihhub 
I've investigated the problem.
The issue happens on a particular map (Terrain wars) where this object (Sirens) has different passability (for both original game and fheroes2). 
OG:
![image](https://user-images.githubusercontent.com/55348946/183645293-221e5850-0a2f-48a2-a6d9-ad1bf2a2deaa.png)
I don't know, how it happened and should we fix it.
But, nevertheless, there're differences in passabilities in the OG and fheroes2 regarding this object.
fheroes2:
![image](https://user-images.githubusercontent.com/55348946/183646785-313f79d4-d292-496c-a2e6-95fc242d69b4.png)

**Comment by ihhub** [comment]
No, we shouldn't fix obvious hacked map format.

**Comment by Branikolog** [comment]
> No, we shouldn't fix obvious hacked map format.

Okay. But nevertheless fheroes2 works differently, comparing to the OG regarding this place. Could you, please, investigate whether these differences could cause some other problems in future?

**Comment by LeHerosInconnu** [comment]
I don't have this scenario either.
